[CSS Zoom] Evaluate container queries against unzoomed sizes.

sammygill opened the pull request:
#### 819df563a4f63736b23e5617bdf7f3320f90bed6
<pre>
[CSS Zoom] Evaluate container queries against unzoomed sizes.
<a href="https://bugs.webkit.org/show_bug.cgi?id=300066">https://bugs.webkit.org/show_bug.cgi?id=300066</a>
<a href="https://rdar.apple.com/161861142">rdar://161861142</a>

Reviewed by Antti Koivisto, Sam Weinig, and Matthieu Dubet.

There was a CSSWG resolution which clarified the behavior between the
zoom property and resolving container queries in terms of what size
values the container query should use. The resolution was to use the
unscaled dimensions of the principle box.

<a href="https://github.com/w3c/csswg-drafts/issues/10268">https://github.com/w3c/csswg-drafts/issues/10268</a>

This patch aligns WebKit with other engines by removing the scale from
the principle box when evaluating a container query since we apply zoom
to the renderer&apos;s sizes during layout. There are two exceptions we use
to decide not to apply this logic:

1.) If the computed value of the propery is a non-fixed value.
Technically the spec only calls out the auto or percent cases but it
seems other engines also do not apply zoom to other non-fixed values
(e.g. intrinsic keywords).

2.) If the RenderStyle associated with the principle box has the
useSVGZoomRulesForLength bit set. This is because we handle zoom
differently since we seem to scale the entire document. This may be
correct but I opt for the safest case here which is to not change
functionality.

Tests: imported/w3c/web-platform-tests/css/css-viewport/zoom/container-queries-block-size.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/container-queries-height.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/container-queries-inline-size.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/reference/container-queries-block-size-ref.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/reference/container-queries-height-ref.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/reference/container-queries-inline-size-ref.html
I also added a few more tests to cover container queries against the
container height and also make sure we are checking the inline/block
dimensions correctly.

Canonical link: <a href="https://commits.webkit.org/301639@main">https://commits.webkit.org/301639@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3853bf1b4d90bb1016ffccb14f3d7895a0f4f2b5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126408 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46055 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/36944 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/133336 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78121 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/6a441266-490b-4457-8044-97b6f9dbf782) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/128279 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46715 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54601 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96224 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/64317 "Passed tests") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129356 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37369 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/113061 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76697 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/750d3a2e-b1e5-4000-ab02-448e5b946d5f) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36259 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/31239 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/76632 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/107177 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31504 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/135862 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/53123 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/40832 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104718 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53596 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/109355 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104417 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26665 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/49878 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28212 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50520 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53034 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/58858 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52334 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55669 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54058 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

mdubet reviewed (approved):
It seems we could factorize there 4 different query into an helper function which takes the computed size and the layout no?

vitorroriz commented on LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/reference/container-queries-block-size-ref.html:
```diff
@@ -0,0 +1,34 @@
+<!doctype html>
+<title>Reference for container-queries-block-size.html</title>
+<link rel="author" title="Chris Harelson" href="mailto:chrishtr@chromium.org">
```
is this correct?

vitorroriz commented on LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/reference/container-queries-height-ref.html:
```diff
@@ -0,0 +1,33 @@
+<!doctype html>
+<title>Reference for container-queries-height.html</title>
+<link rel="author" title="Chris Harelson" href="mailto:chrishtr@chromium.org">
```
ditto

vitorroriz commented on LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/reference/container-queries-inline-size-ref.html:
```diff
@@ -0,0 +1,34 @@
+<!doctype html>
+<title>Reference for container-queries-inline-size.html</title>
+<link rel="author" title="Chris Harelson" href="mailto:chrishtr@chromium.org">
```
ditto

vitorroriz commented on Source/WebCore/css/query/ContainerQueryFeatures.cpp:
```diff
@@ -64,6 +73,7 @@ struct SizeFeatureSchema : public FeatureSchema {
         return evaluate(feature, *renderer, context.conversionData);
     }
 
+
```
nit: unrelated

nullhook reviewed (commented):
this doesn't respect enableEvaluationTimeZoom flag

nt1m commented on Source/WebCore/css/query/ContainerQueryFeatures.cpp:
```diff
@@ -42,6 +42,15 @@ namespace WebCore::CQ {
 
+enum class UsesSVGZoomRulesForLength : bool { No, Yes };
+
+static LayoutUnit unscaledSizeForPrincipleBox(const Style::PreferredSize& computedSize, LayoutUnit usedSize, UsesSVGZoomRulesForLength usesSVGZoomRulesForLength, const Function<float()>& zoomFunction)
```
Why are we not just passing the zoom float directly instead of passing a function to be called?

sammygill replied:
It was to avoid passing a zoom that would go unused but now that I think about it this probably isn't giving us anything beneficial in that regard

nullhook commented on Source/WebCore/css/query/ContainerQueryFeatures.cpp:
```diff
@@ -42,6 +42,15 @@ namespace WebCore::CQ {
+    if (usesSVGZoomRulesForLength == UsesSVGZoomRulesForLength::Yes || computedSize.isAuto() || computedSize.isPercent())
+        return usedSize;
+    return LayoutUnit { usedSize / zoomFunction() };
+}
```
i think you can send the whole style object here instead?

```
static LayoutUnit unscaledSizeForPrincipleBox(const Style::PreferredSize& computedSize, LayoutUnit usedSize, const RenderStyle& style) {
    if (style.useSVGZoomRulesForLength() || ...)
        return usedSize;

    return LayoutUnit { usedSize / style.usedZoom() };
}
```

and then at call-site

`auto width = unscaledSizeForPrincipleBox(style.width(), renderer.contentBoxWidth(), style);`

nullhook replied:
let's make the name of this routine more obvious perhaps `computedSizeWithoutZoom` - because that's what it actually returns.

sammygill replied:
I don't think that's what it is returning

nullhook commented on Source/WebCore/css/query/ContainerQueryFeatures.cpp:
```diff
@@ -93,7 +106,11 @@ struct HeightFeatureSchema : public SizeFeatureSchema {
     {
-        return evaluateLengthFeature(feature, renderer.contentBoxHeight(), conversionData);
+        auto& renderStyle = renderer.style();
+        auto usesSVGZoomRulesForLength = renderStyle.useSVGZoomRulesForLength() ? UsesSVGZoomRulesForLength::Yes : UsesSVGZoomRulesForLength::No;
```
what's the purpose for this indirection? could we just pass RenderStyle& and let the helper function read the flags directly?

sammygill replied:
I personally like to pass in only the information needed for the function in order to do its job. Passing in large objects like RenderStyle would require the function to not only do the work it was designed to, but also try and figure out what data it needs to do that by grabbing things off objects that contain more information than it likely needs.

weinig commented on Source/WebCore/css/query/ContainerQueryFeatures.cpp:
```diff
@@ -42,6 +42,15 @@ namespace WebCore::CQ {
 
 using namespace MQ;
 
+enum class UsesSVGZoomRulesForLength : bool { No, Yes };
```
This should probably be defined in RenderStyleConstants.h and returned from RenderStyle::useSVGZoomRulesForLength().

sammygill replied:
I moved it to RenderStyleConstants.h but refrained from changing RenderStyle::useSVGZoomRulesForLength() to avoid polluting this PR. What do you think about doing that in a follow up?

weinig replied:
Totally fine.

weinig commented on Source/WebCore/css/query/ContainerQueryFeatures.cpp:
```diff
@@ -42,6 +42,15 @@ namespace WebCore::CQ {
+
+static LayoutUnit unscaledSizeForPrincipleBox(const Style::PreferredSize& computedSize, LayoutUnit usedSize, UsesSVGZoomRulesForLength usesSVGZoomRulesForLength, float usedZoom)
+{
+    if (usesSVGZoomRulesForLength == UsesSVGZoomRulesForLength::Yes || computedSize.isAuto() || computedSize.isPercent())
```
This should probably be 

```
if (usesSVGZoomRulesForLength == UsesSVGZoomRulesForLength::Yes || !computedSize.isFixed())
```

I don't think undoing zoom is wanted when any of the other intrinsic keywords or calc() are used.

sammygill replied:
Changed! Yeah I wasn't sure since the spec calls out only auto and percent for some reason but it seems like other engines also do not apply zoom to those cases which makes sense to me

sammygill committed:
[CSS Zoom] Evaluate container queries against unzoomed sizes.
https://bugs.webkit.org/show_bug.cgi?id=300066
rdar://161861142

Reviewed by Antti Koivisto, Sam Weinig, and Matthieu Dubet.

There was a CSSWG resolution which clarified the behavior between the
zoom property and resolving container queries in terms of what size
values the container query should use. The resolution was to use the
unscaled dimensions of the principle box.

https://github.com/w3c/csswg-drafts/issues/10268

This patch aligns WebKit with other engines by removing the scale from
the principle box when evaluating a container query since we apply zoom
to the renderer's sizes during layout. There are two exceptions we use
to decide not to apply this logic:

1.) If the computed value of the propery is a non-fixed value.
Technically the spec only calls out the auto or percent cases but it
seems other engines also do not apply zoom to other non-fixed values
(e.g. intrinsic keywords).

2.) If the RenderStyle associated with the principle box has the
useSVGZoomRulesForLength bit set. This is because we handle zoom
differently since we seem to scale the entire document. This may be
correct but I opt for the safest case here which is to not change
functionality.

Tests: imported/w3c/web-platform-tests/css/css-viewport/zoom/container-queries-block-size.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/container-queries-height.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/container-queries-inline-size.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/reference/container-queries-block-size-ref.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/reference/container-queries-height-ref.html
       imported/w3c/web-platform-tests/css/css-viewport/zoom/reference/container-queries-inline-size-ref.html
I also added a few more tests to cover container queries against the
container height and also make sure we are checking the inline/block
dimensions correctly.

Canonical link: https://commits.webkit.org/301639@main
